File: java-breeds/app/src/main/java/es/tipolisto/breeds/utils/MediaPlayerClient.java
package es.tipolisto.breeds.utils;

import android.content.Context;
import android.media.AudioManager;
import android.media.MediaPlayer;


import es.tipolisto.breeds.R;
import es.tipolisto.breeds.data.preferences.PreferencesManagaer;


public class MediaPlayerClient {
    //La otra clase para reproducir audi se llama AudioTrack: https://developer.android.com/reference/android/media/AudioTrack?hl=es-419
    private MediaPlayer mediaPlayer;
    private Context context;
    private PreferencesManagaer preferencesManagaer;

    public MediaPlayerClient(Context context){
        mediaPlayer=new MediaPlayer();
        this.context=context;
        preferencesManagaer=new PreferencesManagaer(context);
        //Es posible crear un efecto fade in y fade out con VolumeShaper pero el dispositivo tiene que ser mayor de android 8.0 api 26
        /*if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            VolumeShaper.Configuration config =
                    new VolumeShaper.Configuration.Builder()
                            .setDuration(3000)
                            .setCurve(new float[] {0.f, 1.f}, new float[] {0.f, 1.f})
                            .setInterpolatorType(VolumeShaper.Configuration.INTERPOLATOR_TYPE_STEP)
                            .build();
            VolumeShaper volumeShaper= mediaPlayer.createVolumeShaper(config);;
            mediaPlayer.setNextMediaPlayer(mediaPlayer);
        }*/

    }



    /*public static MediaPlayerClient getInstance(){
        if(instance==null){
            instance=new MediaPlayerClient();
        }
        return instance;
    }*/

    public void playIntro(){
        if (!preferencesManagaer.getMusicOnOff()) return;
        mediaPlayer = MediaPlayer.create(context, R.raw.intro);
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mediaPlayer.start();
        mediaPlayer.setLooping(true);




    }
    public void playInGame(){
        if (!preferencesManagaer.getMusicOnOff()) return;
        if (mediaPlayer.isPlaying()) {
            mediaPlayer.stop();
            mediaPlayer.release();
        }
        mediaPlayer = MediaPlayer.create(context, R.raw.ingame);
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mediaPlayer.start();
        mediaPlayer.setLooping(true);
    }
    /*
    public void playGameOver(Context context){
        mediaPlayer = MediaPlayer.create(context, R.raw.gameover);
        mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
        mediaPlayer.start();
    }*/

    public void playSound(String sound){
        if (!preferencesManagaer.getMusicOnOff()) return;
        mediaPlayer.stop();
        mediaPlayer.release();
        if (sound.equals("button")) mediaPlayer = MediaPlayer.create(context, R.raw.clickbutton);
        if (sound.equals("success")) mediaPlayer = MediaPlayer.create(context, R.raw.success);
        if (sound.equals("failure")) mediaPlayer = MediaPlayer.create(context, R.raw.failure);
        mediaPlayer.start();
    }
    public void stopSound(){
        if(mediaPlayer.isPlaying()) mediaPlayer.stop();
    }
    public void releaseSound(){
        if (!preferencesManagaer.getMusicOnOff()) return;
        if(mediaPlayer.isPlaying())
            mediaPlayer.stop();
        mediaPlayer.release();
        mediaPlayer = null;
    }
}

File: kotlin-breeds/app/src/main/java/es/tipolisto/breeds/utils/MediaPlayerClient.kt
package es.tipolisto.breeds.utils

import android.content.Context
import android.media.MediaPlayer
import android.util.Log
import es.tipolisto.breeds.R


class MediaPlayerClient (private val context: Context) {
    private var inGameMusic:MediaPlayer
    private var menuMusic:MediaPlayer
    private var endMusic:MediaPlayer
    private var effects:MediaPlayer
    init {
        inGameMusic = MediaPlayer.create(context, R.raw.ingame)
        menuMusic= MediaPlayer.create(context, R.raw.intro)
        endMusic= MediaPlayer.create(context, R.raw.gameover)
        effects= MediaPlayer.create(context, R.raw.clickbutton)
        inGameMusic.isLooping=true
        menuMusic.isLooping=true
    }
    fun playSound(effectsType: AudioEffectsType){
        var effects:MediaPlayer
        when(effectsType.name){
            "click"-> effects=MediaPlayer.create(context, R.raw.clickbutton)
            "success"-> effects = MediaPlayer.create(context, R.raw.success)
            "failure"-> effects = MediaPlayer.create(context, R.raw.failure)
            else->{
                effects = MediaPlayer.create(context, R.raw.failure)
            }
        }
        effects.start()
    }
    /*fun getInstance(): MediaPlayerClient {
        if (instance == null) {
            synchronized(this) {
                if (instance == null) {
                    instance = MediaPlayerClient(context)
                }
            }
        }
        return instance!!
    }*/
    fun playInGameMusic(){
        inGameMusic = MediaPlayer.create(context, R.raw.ingame)
        if(!inGameMusic.isPlaying)inGameMusic.start()
    }

    fun stopInGameMusic(){
        if(inGameMusic.isPlaying)inGameMusic.stop()
    }

    fun playMenuMusic(){
        menuMusic= MediaPlayer.create(context, R.raw.intro)
        //Log.d("TAG","MediaPlayerClient dice: Hemos entrado en reproducir musica del menu")
        //menuMusic= MediaPlayer.create(context, R.raw.intro)
        if(!menuMusic.isPlaying){
            //menuMusic.reset()
            menuMusic.start()
            //Log.d("TAG","MediaPlayerClient dice: reproduciendo musica menu")
        }
    }

    fun stopMenuMusic(){
        //Log.d("TAG","MediaPlayerClient dice: Hemos entrado en parar musica menu")
        if(menuMusic.isPlaying){
            menuMusic.stop()
            //Log.d("TAG","MediaPlayerClient dice: musica menu parada")
        }
    }
    fun clickAudio(){
        effects.start()
    }

    /*fun updateAudio() {
        var musicOnOff= PreferenceManager.readPreferenceMusicOnOff(contextFromClient!!)
        //val mediaPlayer=MediaPlayer.create(context,R.raw.intro)
        if(musicOnOff){
            if(!mediaPlayer!!.isPlaying)mediaPlayer!!.start()
            else
                if(mediaPlayer!!.isPlaying)mediaPlayer!!.stop()
        }
    }*/


}


